recent_topics: Add Private message to recent_topics.

madrix01 opened the pull request:
We add Private message to the Recent topics to make it an
all-encompassing overview of everything that's happening for a user.

![image](https://user-images.githubusercontent.com/51414879/165738465-570a47ad-fd75-42ec-94d2-f32cee70f747.png)
Fixes #19449

amanagr commented on static/js/recent_topics_data.js:
```diff
@@ -6,16 +6,18 @@ export const topics = new Map(); // Key is stream-id:topic.
     // has changed or not.
-    if (msg.type !== "stream") {
-        // We don't process private messages yet.
+    if (msg.type !== "stream" && msg.type !== "private") {
```
We don't need this condition anymore. Message is of only two types.

amanagr commented on static/js/recent_topics_data.js:
```diff
@@ -6,16 +6,18 @@ export const topics = new Map(); // Key is stream-id:topic.
+    // Initialize topic and pm data
+    // Key for private message is the user id's
+    // to whom the message is begin sent.
+    const key = msg.type === "stream" ? get_topic_key(msg.stream_id, msg.topic) : msg.to_user_ids;
```
We should extract this to a function `get_key_from_message`.

madrix01 replied:
@amanagr is it okay to set the `pm_key` as `to_user_ids` ?

amanagr replied:
It is fine but `recipient_id` is better. So, we should use `recipient_id` here, thanks for asking.

amanagr commented on static/js/recent_topics_ui.js:
```diff
@@ -255,7 +256,6 @@ function format_topic(topic_data) {
     const all_senders = recent_senders.get_topic_recent_senders(stream_id, topic);
     const senders = all_senders.slice(-MAX_AVATAR);
     const senders_info = people.sender_info_for_recent_topics_row(senders);
-
     // Collect extra senders fullname for tooltip.
     const extra_sender_ids = all_senders.slice(0, -MAX_AVATAR);
     const displayed_other_senders = extra_sender_ids.slice(-MAX_EXTRA_SENDERS);
```
Why remove black new lines? They help in reading code sometimes.

madrix01 replied:
It must have been by mistake. Resolved!!

amanagr commented on static/js/recent_topics_ui.js:
```diff
@@ -303,6 +302,61 @@ function format_topic(topic_data) {
     };
 }
 
+function format_pm(pm_data) {
```
Let's extract a function to return the common details between `format_pm` and `format_topic` instead of repeating all the code here.

amanagr commented on static/js/recent_topics_ui.js:
```diff
@@ -344,31 +398,37 @@ export function update_topics_of_deleted_message_ids(message_ids) {
     const msg = message_store.get(topic_data.last_msg_id);
     const sub = sub_store.get(msg.stream_id);
-
-    if (sub === undefined || !sub.subscribed) {
+    if ((sub === undefined || !sub.subscribed) && topic_data.type !== "private") {
         // Never try to process deactivated & unsubscribed stream msgs.
         return true;
     }
```
This needs a condition to check the message is a stream message and remove the ` topic_data.type !== "private"` condition from here since we are still processing `unread` private messages.

madrix01 replied:
I have changed it to `topic_data.type === "stream"`. Unread PM's are still processed with this.

amanagr commented on static/js/recent_topics_ui.js:
```diff
@@ -344,31 +398,37 @@ export function update_topics_of_deleted_message_ids(message_ids) {
 
     if (filters.has("unread")) {
-        const unreadCount = unread.num_unread_for_topic(msg.stream_id, msg.topic);
+        let unreadCount;
+        if (msg.type === "private") {
+            unreadCount = unread.num_unread_for_person(msg.to_user_ids);
+        } else if (msg.type === "stream") {
+            unreadCount = unread.num_unread_for_topic(msg.stream_id, msg.topic);
+        }
```
Extract this into `get_thread_unread_count_from_message` and use it in both extracted `format_pm/topic` function above and here.

amanagr commented:
@madrix01 thanks for working on this! Left some comments.

madrix01 commented:
@amanagr I have made most of the changes. I'm facing problem with keyboard navigation. 
![Recent topics - Zulip Dev - Zulip](https://user-images.githubusercontent.com/51414879/152563795-91b0cf70-2edb-45d1-b4be-13211bce3201.gif)

madrix01 commented:
It got solved!!

timabbott commented on frontend_tests/node_tests/recent_topics.js:
```diff
@@ -707,17 +731,6 @@ test("basic assertions", ({override_rewire, mock_template}) => {
-    assert.equal(
-        Array.from(all_topics.keys()).toString(),
-        "4:topic-10,1:topic-7,1:topic-6,1:topic-5,1:topic-4,1:topic-3,1:topic-2,1:topic-1",
-    );
```
We should replace this test with something verifying the processing of private messages :)

madrix01 commented:
@amanagr I have update the commit with private tests as well, can you please have a look at it?

amanagr commented:
@madrix01 did you lose all your changes 😅 ? I only see some relevant code here. `git reflog` can help you recover those changes.

madrix01 commented:
@amanagr no, they are all present in this [commit](https://github.com/zulip/zulip/commit/3559a6e41f09148f28bb42edbe0e20c9439ab2f1) as time asked me to break the commit into small changes, so it would be easy to review the code and add tests. [Discussed here](https://chat.zulip.org/#narrow/stream/101-design/topic/Recent.20topics.20UI/near/1324360), all the changes you have suggested are present in the above mentioned commit.

amanagr commented:
Why is that commit not part of this PR?

madrix01 commented:
@amanagr It was the part of the pr. I rebased this pr, as i was to break single commit into small commits.

amanagr commented:
So, did you open a PR for the rest of the code? I should be able to test the complete thing. :) It fine to just push remaining commits in this PR too if you haven't done so.

madrix01 commented:
@amanagr @timabbott  apologies for the delay. I have update all the code here including the ui.

timabbott commented:
@madrix01 you'll need to debug the node tests for this before we can merge it.

madrix01 commented:
@timabbott I have tweaked the tests current tests, they work with full coverage, but special tests for private messages are yet to be added.

timabbott commented:
This needs a rebase for our recent topics PRs merged recently, but should otherwise be good to test-deploy on chat.zulip.org.

timabbott commented:
Also, is the first commit something we could merge independently of the second one to reduce the size of this PR and thus merge conflict risk?  I think we might need some checks to prevent trying to display those without `type="stream"` in order to do that.

madrix01 commented:
@timabbott 

> Also, is the first commit something we could merge independently of the second one to reduce the size of this PR and thus merge conflict risk? I think we might need some checks to prevent trying to display those without `type="stream"` in order to do that.

Yes this commit will only allow message with all the type to pass in `process_messages()` in `recent_topics_data.js` but PMs will not be rendered in the UI, there is a check for filtering out messages with only `type === "stream"`

alya commented:
I just pulled down the PR, and it's not quite ready for test deployment. I'll post my comments here.

alya commented:
The unread counts need to be aligned properly:
<img width="854" alt="Screen Shot 2022-03-08 at 12 00 04 PM" src="https://user-images.githubusercontent.com/2090066/157315474-7a477d8a-e931-411d-872f-4b3a9fd6e03d.png">

alya commented:
The icons (envelope and people) are too dark in light mode; I think they'd look good matching the background of the unread counts.

The colors seem reasonable in dark mode, but I'm curious how they were selected.

alya commented:
I'm also getting this error if I click around:
![Screen Shot 2022-03-08 at 12 13 28 PM](https://user-images.githubusercontent.com/2090066/157317793-a8971479-fd63-497d-9761-3210085c7dc3.png)

alya commented:
Otherwise looks good so far; I can test more when the error is no longer popping up.

madrix01 commented:
@alya I'll try to solve it and update the PR ASAP. Thank you!

madrix01 commented:
> The unread counts need to be aligned properly: <img alt="Screen Shot 2022-03-08 at 12 00 04 PM" width="854" src="https://user-images.githubusercontent.com/2090066/157315474-7a477d8a-e931-411d-872f-4b3a9fd6e03d.png">

Absence of mark as read and mute topic icon was causing this UI bugs, I have added margin to the right to match the unread in the topics.
![image](https://user-images.githubusercontent.com/51414879/157435532-7e006236-3648-4b2b-b52c-3ff3fe42ddf9.png)

madrix01 commented:
> I'm also getting this error if I click around: ![Screen Shot 2022-03-08 at 12 13 28 PM](https://user-images.githubusercontent.com/2090066/157317793-a8971479-fd63-497d-9761-3210085c7dc3.png)

Can you show me the steps to generate this error?

alya commented:
Hm, I'm not getting the error anymore, so maybe my dev environment somehow got into a weird state. Could you push your other fixes? Thanks!

madrix01 commented:
@alya @timabbott I have pushed the following changes
- made pm row unread counts align with topic row
- fixed bugs with keyboard navigation as there are only 2 columns with PM row and 4 with columns row
- added function to differentiate between Pm and topic row.

Will add more ui changes soon.

Thank you!

madrix01 commented:
@alya 
![image](https://user-images.githubusercontent.com/51414879/157832814-a2f94863-415d-44fd-9daf-105949bbb165.png)
Icons in light mode

![image](https://user-images.githubusercontent.com/51414879/157840067-f20f9988-6ee1-4961-ac29-3e35a7dc060a.png)
Icons in dark mode

madrix01 commented:
@alya I have updated the PR with the changes you suggested can you please have a look at it?
Thank you !!

alya commented:
Thanks! I'm still seeing bad unread counter alignment in narrow screen mode:
![Screen Shot 2022-03-15 at 10 43 30 PM](https://user-images.githubusercontent.com/2090066/158524389-f0a85ddf-5a07-4f88-9cf7-bcc2ab3877e4.png)

alya commented:
@madrix01  Could you please go through the manual review checklist at https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#manual-testing, as there are a lot of opportunities for UI glitches on this one?

madrix01 commented:
Ok, I'll fix that, Are you facing any bug or glitches with the keyboard navigation ?

alya commented:
I haven't tested keyboard navigation.

timabbott commented on static/templates/recent_pm_row.hbs:
```diff
@@ -0,0 +1,49 @@
+        <div class="flex_container">
+            <div class="left_part recent_topics_focusable">
+                <span class="fa fa-envelope"></span>
+                <a href="{{pm_url}}">Private messages</a>
```
All strings should be tagged for translation. 

https://zulip.readthedocs.io/en/latest/translating/internationalization.html

timabbott commented on static/templates/recent_pm_row.hbs:
```diff
@@ -0,0 +1,49 @@
+                </li>
+            {{/each}}
+        </ul>
+    </td>
```
What's your thinking on the approach of adding a new template here vs. adding conditionals in `static/templates/recent_topic_row.hbs`?  It seems like several blocks, including this one and the one below, could be identical or nearly identical, and it'd potentially help a lot for readability/maintainability to share code.

madrix01 replied:
@timabbott Main concerns for creating new template was introduction of lots of conditionals in `recent_topic_row.hbs` for icons, titles etc. which I think it would clutter the template, and some variables name wouldn't match their use, that could be confusing, or maybe we can use common names such as `thread_key` in place of `topic_key` and `pm_key`.

timabbott replied:
Yeah, I think `conversation_key` would be a good common variable name -- we can do a preparatory refactoring commit to migrates to that.  

I think it's OK for moderately size blocks to be inside conditions -- it's just silly to copy the entire file when large sections are identical. (This often leads to problems where we fix a bug in only one copy).

timabbott commented on static/styles/recent_topics.css:
```diff
@@ -262,6 +287,14 @@
             width: 40%;
+
+            .line_clamp {
+                /* stylelint-disable-next-line value-no-vendor-prefix */
```
We should include an explanation of why we think the rule should not apply here when doing these overrides.

timabbott commented on static/styles/recent_topics.css:
```diff
@@ -297,6 +330,10 @@
             }
+
+            .unread_count_pm {
+                margin-right: 53px;
```
How is this 53px determined? It's possible we want to do some math with variables like we have in `left-sidebar.css` to compute this value... magic numbers like 53px are rarely a good way to do things.

timabbott commented on static/styles/recent_topics.css:
```diff
@@ -131,6 +144,18 @@
         }
 
+        .unread_count_pm {
+            margin-right: 46px;
```
Same question with this 46px; I don't understand how it related to the 53px below either.

timabbott commented on static/styles/recent_topics.css:
```diff
@@ -131,6 +144,18 @@
+            min-width: 7px;
+            height: 7px;
+            margin-left: 20px;
+            margin-right: -7px;
```
Can you add a comment explaining the intent of this block?  Negative margins are generally a bad pattern, and I wonder if we could do something more declarative with flex or another modern CSS feature here.

madrix01 commented:
@timabbott I have updated pull request with changes you suggested and separated one preparatory commit. But it is failing test coverage, I think mentioned line doesn't have any major code changes and there isn't any file mentioned in the error message.

timabbott commented:
The error is this:
```
ERROR: frontend_tests/node_tests/recent_topics.js no longer has complete node test coverage
  Lines missing coverage: 143, 145, 147
```

This means you added some code in that test, perhaps a function, that never gets called. Sometimes that's a bug; if it's intentional, you can add a comment to disable coverage (I think there's a few scattered through the node tests, but forgot the syntax).

madrix01 commented:
@timabbott I have figured the line causing error and resolved, I think it's ready for review. Thank you!

alya commented:
Nice, I think we're getting there! A few comments from playing with the PR:

1. Having the filter be called "Private messages" doesn't seem right, as it seems to imply that you would only be viewing PMs. I think we should rename it to "Include PMs", and move to be between "All" and "Include muted".
2. The "Private messages" link in the left column goes to just the thread, not all PMs. While we may want to change how this works, for now, I think to be consistent with streams, clicking on the "Private messages" text should go to the all PMs view. @timabbott does that sound right?
3. Is it intentional that when there is no unread counter, the names spill into the right-hand columns? I think it would look better if they did not.

<img width="699" alt="Screen Shot 2022-04-27 at 2 48 50 PM" src="https://user-images.githubusercontent.com/2090066/165636971-3ec43055-7a2e-4b4e-8f43-638661f2c6f8.png">

madrix01 commented:
@alya 
> Is it intentional that when there is no unread counter, the names spill into the right-hand columns? I think it would look better if they did not.

No, it's not intentional. This happens because we can't break the names after the maximum width. 
This happens with long topic names too 
![image](https://user-images.githubusercontent.com/51414879/165735425-46152b83-069f-4e00-a050-6111951977df.png)

timabbott commented:
> The "Private messages" link in the left column goes to just the thread, not all PMs. While we may want to change how this works, for now, I think to be consistent with streams, clicking on the "Private messages" text should go to the all PMs view. @timabbott does that sound right?

Yep, this sounds right!

madrix01 commented:
And we can replace private message envelope with user presence icon.

madrix01 commented:
@timabbott @alya I have made this [prototype](https://design.penpot.app/#/view/34b00100-7dc2-11ec-a93c-e5d362d27ee9?page-id=34b00101-7dc2-11ec-a93c-e5d362d27ee9&section=interactions&index=0&share-id=989f3cc0-d6b5-11ec-ab9e-75f6b8214121) as per Tim's suggestion.

alya commented:
> No, it's not intentional. This happens because we can't break the names after the maximum width.
This happens with long topic names too

I see. Would you like to fix this in general as part of this PR, or should I file a separate issue?

alya commented:
> @timabbott @alya I have made this [prototype](https://design.penpot.app/#/view/34b00100-7dc2-11ec-a93c-e5d362d27ee9?page-id=34b00101-7dc2-11ec-a93c-e5d362d27ee9&section=interactions&index=0&share-id=989f3cc0-d6b5-11ec-ab9e-75f6b8214121) as per Tim's suggestion.

I think the suggestion was to use *both* the stream and topic columns, rather than moving the names to *one* of the other columns. In any case, we can keep it as-is for now, I think.

madrix01 commented:
> I see. Would you like to fix this in general as part of this PR, or should I file a separate issue?

I'll make separate commit for this one in this PR.

madrix01 commented:
@alya I made PR #22180 to prevent spilling Topic names into unread counter column.

alya commented:
I think this one still needs to be addressed:

> Having the filter be called "Private messages" doesn't seem right, as it seems to imply that you would only be viewing PMs. I think we should rename it to "Include PMs", and move to be between "All" and "Include muted".

alya commented:
I also just noticed that the envelope looks a little low:
![Screen Shot 2022-06-03 at 12 37 53 AM](https://user-images.githubusercontent.com/2090066/171809894-2eb492bc-f5f4-411a-9634-2c0cf3fb8792.png)

It should be aligned at the bottom with the "Private messages" text, similarly to the `#` symbol for streams.

madrix01 commented:
@alya I'll rebase and push the change ASAP.

alya commented:
Cool, thanks! Please post screenshots addressing both of the comments above when ready.

madrix01 commented:
@alya I have changed the filter. The alignment of both hash and envelope icon is same, i.e in center, I think it seems of mainly due to 'P' and the icon itself

alya commented:
Hm, do you mean that the icons are vertically centered? Maybe they should both be bottom-aligned then... Whatever the implementation, I think the current visuals don't look right.

madrix01 commented:
![image](https://user-images.githubusercontent.com/51414879/172095009-d25b8374-ccc8-45c9-b376-131f1ee5ab65.png)
@alya Does this look correct? I have updated PR.

alya commented:
Yeah, that looks good to me, thanks!

alya commented:
What's the status of this PR at this point? Does it need a review from @amanagr ?

madrix01 commented:
@alya The keyboard navigation is yet to be implemented with the new unread counters. I'm first trying to implement it on #22219.

timabbott commented:
@madrix01 it'd be great if you can prioritize fixing the keyboard UI on this; I think it's otherwise deployable on chat.zulip.org, and from looking at the code, it seems like the keyboard UI might not be hard to fix?

madrix01 commented:
@timabbott I have rebased the PR earlier to make it work with the Unread Counter as button. I'll fix this puppeteer error.

madrix01 commented:
@timabbott @amanagr I have fixed the puppeteer test as well. The keyboard navigation is also working fine.

timabbott commented:
Cool, I'm test-deploying this on chat.zulip.org for feedback.

timabbott commented:
@madrix01 any update on when you'll been able to fix the bugs reported in this PR? This needs to be merged soon if it's going to make it into 6.0.

madrix01 commented:
I'll fix it as soon as possible!!

madrix01 commented:
@timabbott @alya I have pushed the code with bugs fixes can you please have a look ?

timabbott commented:
Looks plausible; I'm updating the deployment.

timabbott commented:
@madrix01 do you know whether the CI failure is related to this PR? If not, please rebase; otherwise, we need to debug that in order to proceed with work on integrating this.

madrix01 commented:
@timabbott This error is due to timeout, all test passed when I ran same tests on my local environment.

timabbott commented:
Updating the chat.zulip.org deployment of this.

timabbott commented on static/templates/recent_topic_row.hbs:
```diff
@@ -1,18 +1,43 @@
             </div>
             <div class="right_part">
+                {{#if is_private}}
+                <span class="unread_count unread_count_pm {{#unless unread_count}}unread_hidden{{/unless}}">{{unread_count}}</span>
```
Why is this if/else outside the `recent_topics_actions` block? Oh, I guess we have this existing weird setup -- the name of `recent_topics_actions` suggests there is a single div, but there's actually just one.

And I guess it's intentional that you can't mark as read from this UI, but that seems like something we'd want to change in a follow-up?

madrix01 replied:
Mark as read is possible from this UI, it just takes some time to rerender and remove the unread counter.

timabbott replied:
No, if you hover over a private message count, one can't mark it as read here. I think that's probably a bug?

timabbott commented on static/templates/recent_topic_row.hbs:
```diff
@@ -1,18 +1,43 @@
+<tr id="recent_conversation:{{conversation_key}}" {{#if unread_count}}class="unread_topic"{{/if}} data-unread-count="{{unread_count}}" data-muted="{{muted}}" data-participated="{{participated}}">
     <td class="recent_topic_stream">
+        {{#if is_private}}
+        <div class="flex_container flex_container_pm">
```
Why do we have this containing `div` with the `flex_container` class only for private messages? I'm wondering if we can shrink the region that has this if/else block, and/or just make the markup more similar, which in turn would make reading the CSS easier.

(I did notice that the topic block does have both stream and PM versions inside a shared `flex_container` div).

madrix01 replied:
In `Private message` row we need to show the presence/group indicator for PMs, for this reason I have used the `flex_container` for only PMs. I'll try to shrink it.

timabbott commented on static/styles/recent_topics.css:
```diff
@@ -287,11 +322,21 @@
+            .line_clamp {
+                /* As display: -webkit-box is not valid for */
+                /* display property in stylelint library */
+                /* stylelint-disable-next-line value-no-vendor-prefix */
```
This explanation is not particularly satisfying. I believe StyleLint bans using prefixes like `-webkit-box` because they are not going to work on Firefox; see: https://caniuse.com/?search=webkit-box

You can test on Firefox to see if this is indeed broken there, but generally our preference will be to use a different CSS mechanism that is part of the standard.

madrix01 replied:
I have tested with firefox and is working as expected.

madrix01 replied:
@timabbott I have tested both on firefox and chrome, line clamping using webkit is working fine. Can we use it in our final versions with suitable comments?

timabbott replied:
I'll change the comment and integrate this, but please start a #frontend thread about whether there's a better way to do line clamping without this `webkit-box` feature that is webkit-specific.

madrix01 replied:
Yes, I will start the conversation there.

timabbott commented on static/styles/recent_topics.css:
```diff
@@ -148,10 +162,31 @@
+            height: 7px;
+            margin-left: 8px;
+            top: 0;
+        }
```
How are these numbers picked? And how do they related to the styles in `user_circles.css`?

madrix01 replied:
@timabbott The size of the `user_circle` is kept constant with the user_circle in left sidebar. But it was looking weird so I decreased the size by 1px.

timabbott commented on static/styles/dark_theme.css:
```diff
@@ -635,6 +635,11 @@ body.dark-theme {
+
+        .fa-envelope,
+        .fa-group {
+            color: hsl(221, 13%, 71%);
```
This color does not appear anywhere else in the app; how did you pick it? Could the colors for these icons be represented as just an `opacity` on the default text color for the light/dark themes?

madrix01 replied:
@timabbott I have converted the color of both the icons to `opacity` in dark as well as light mode.

timabbott commented on static/js/unread.js:
```diff
@@ -667,6 +667,13 @@ export function get_missing_topics(opts) {
     return unread_topic_counter.get_missing_topics(opts);
 }
 
+export function get_thread_unread_count_from_message(msg) {
```
Let's call this `message_to_conversation_unread_count(msg)`; I'd also prefer to have it live in `recent_topics_ui.js` rather than here, since this file tries to avoid dealing with full message objects.

timabbott commented on static/js/compose_closed_ui.js:
```diff
@@ -44,6 +44,8 @@ export function get_recipient_label(message) {
         } else if (message.display_reply_to) {
             return message.display_reply_to;
+        } else if (message.recipients) {
+            return message.recipients;
```
Can you explain this change? It seems unrelated to the rest of the pull request and isn't called by the new code.

I also don't expect a real message to fail to have a `display_reply_to`...

Now, it seems very possible we should have an independent commit to tweak this function to have an explicit code path for recent topics; I think at present it probably always returns `""`, but it might be better to make that explicit with a check at the top of the function -- we'll likely want the same for the inbox view too when we do that.

madrix01 replied:
@timabbott This is for the closed composed ui recipient change along with the change in the recent topics row. 
On executing the `set_table_focus`, we call `compose_closed_ui.update_reply_recipient_label(message);` which in turn calls the `get_recipient_label` in `compose_closed_ui`.
https://github.com/zulip/zulip/blob/686bdc23cd78708a688718882d21d7e1797bee72/static/js/recent_topics_ui.js#L188-L201

And the `message` object is obtained from the above code, which only has `stream` and `topic` or `recipients`, the `display_reply_to`

timabbott replied:
OK -- let's add a comment noting that this case is for when this function is passed a recent topics private message row. That will make it a lot easier to understand :).

But I also wonder -- isn't this fake `message` object only used for that one function? 

My read at this point is that we should probably just add a prep commit to `compose_closed_ui.js` that changes the interface between these two modules, and just pass a `label` parameter that's constructed in this function to `update_reply_recipient_label` here. The name `message` is confusing, since it's not a message object we're passing, and we don't seem to be materially sharing code usefully with `get_recipient_label`.

timabbott replied:
I did a bit more reading and think this is a mess that's a bit out of scope here; I'll add some comments and we can open a follow-up issue: https://github.com/zulip/zulip/issues/23250

timabbott commented on static/js/recent_topics_ui.js:
```diff
@@ -116,6 +124,35 @@ function is_table_focused() {
+    // So we use CSS method for finding row type until topics_widget gets initialized.
+    if (!topics_widget) {
+        const $topic_rows = $("#recent_topics_table table tbody tr");
+        const $topic_row = $topic_rows.eq(row);
```
I don't understand the `.find` here; how is this different from just `const $topic_row = row`?

timabbott commented on static/js/recent_topics_ui.js:
```diff
@@ -116,6 +124,35 @@ function is_table_focused() {
+        const $topic_rows = $("#recent_topics_table table tbody tr");
+        const $topic_row = $topic_rows.eq(row);
+        const type = $topic_row.find(".recent_topic_stream a").text();
+        if (type === "Private messages") {
```
This logic will fail once this widget is internationalized. You could check a CSS class instead, but I also wonder if we could instead just have this function not be called before the `topics_widget` is initialized and delete this block?

The comment doesn't super make sense to me; how is `get_current_list` related to this function?

madrix01 replied:
I have added the `data-private` attribute in the `tr` tag which would not pose any problem with internalization.

timabbott replied:
OK, this is improved; but I still don't understand when this function will be called before `topics_widget` is initialized; can you explain that?

madrix01 replied:
@timabbott The bug faced here was that on reload of page, using key board navigation on first few seconds after the reload returned `topics_widget.get_current_list()` undefined.

This is due to the fact that `complete_rerender` calls `revive_current_focus` as `callback_after_render` and `revive_current_focus` calls `set_table_focus` which calls `get_row_type` which uses `get_current_list`

timabbott replied:
OK can you open an issue with a title like "Refactor recent_topics_ui.get_row_type to not explicitly depend on topics_widget having been rendered" with a clear description of that sequence (spelled out a bit more for someone not looking at the code) and reproduction steps for the bug one would have if we removed this block? 

This logic is somewhat brittle but I think we can fix it as a follow-up issue.

timabbott commented on static/js/recent_topics_ui.js:
```diff
@@ -147,25 +184,34 @@ function set_table_focus(row, col, using_keyboard) {
 }
 
 export function get_focused_row_message() {
+    const recent_conversation_id_prefix_len = "recent_conversation:".length;
```
I'd define `const recent_conversation_key_prefix = "recent_conversation:"` at the top of the file, and use that variable both in the `.slice` below and in the related code that generates this prefix.

timabbott commented on static/js/recent_topics_ui.js:
```diff
@@ -252,38 +298,83 @@ export function process_messages(messages) {
-        return {};
-    }
-    const topic = last_msg.topic;
+function format_message(msg_data) {
```
Let's call this `format_conversation(conversation_data) {`; we're doing computations around a conversation, not a message, here.

madrix01 replied:
I have changed the function name to `format_conversation(conversation_data)` and `msg_info` to `context`.

timabbott commented on static/js/recent_topics_ui.js:
```diff
@@ -252,38 +298,83 @@ export function process_messages(messages) {
+        // Collect extra sender fullname for tooltip
+        extra_sender_ids = all_senders.slice(0, -MAX_AVATAR);
+        const displayed_other_senders = extra_sender_ids.slice(-MAX_EXTRA_SENDERS);
+        displayed_other_names = people.get_display_full_names(displayed_other_senders.reverse());
```
I think you can probably deduplicate most of the ~6 lines here for populating the `senders` / etc. fields; only the code to compute `all_senders` and `senders` appears to be different?

timabbott commented on static/js/recent_topics_ui.js:
```diff
@@ -498,26 +574,43 @@ export function update_filters_view() {
+    if (a.toLowerCase() > b.toLowerCase()) {
         return 1;
-    } else if (a_stream === b_stream) {
+    } else if (a === b) {
```
This sorting algorithm is likely wrong; you can't do this comparison without `toLowerCase` when the above one was done with it.

`return util.strcmp(a.toLowerCase(), b.toLowerCase())` might be what you want to use instead of this function anyway.

timabbott commented:
@madrix01 when do you expect to have an updated version of this ready?

madrix01 commented:
@timabbott Today itself. Just a few things remaining. sorry for delay.

timabbott commented on static/js/recent_topics_ui.js:
```diff
@@ -498,26 +574,43 @@ export function update_filters_view() {
+        return sort_comparator(a_msg.display_reply_to, b_msg.topic);
     }
-    return -1;
+    return sort_comparator(a_msg.topic, b_msg.display_reply_to);
```
OK; I'm not sure I understand the bug though.

timabbott replied:
Testing, all that I've observed is that resolved topics are sorted alphabetically after everything, presumably because the unicode check character is larger than the alphabet characters. This seems fine?

madrix01 replied:
@timabbott When using `util.strcmp` for comparing the resolved topics appear at top of the list in sorted manner
![image](https://user-images.githubusercontent.com/51414879/195973937-fad47b68-04be-4690-9f55-ee9b84099c5b.png)
And then comes the unsorted topics (in sorted manner). Is this the expected behaviour?

timabbott replied:
OK yeah the question was unclear. I think what you're saying is replacing the `toLowerCase` comparisons we have with `util.strcmp` would cause resolved topics to be sorted first, which I think I agree we probably don't want. So I think we can stick with what you have in this PR.

timabbott commented:
I pushed back here fixing the test failure and adding a few comments. There's still a couple things I don't understand above.

```
diff --git a/static/js/compose_closed_ui.js b/static/js/compose_closed_ui.js
index cbbdd4f683..dcbd7207c5 100644
--- a/static/js/compose_closed_ui.js
+++ b/static/js/compose_closed_ui.js
@@ -9,6 +9,11 @@ import * as people from "./people";
 import * as recent_topics_util from "./recent_topics_util";
 
 export function get_recipient_label(message) {
+    // TODO: This code path is bit of a type-checking disaster; we mix
+    // actual message objects with fake objects containing just a
+    // couple fields, both those constructed here and potentially
+    // passed in.
+
     if (message === undefined) {
         if (message_lists.current.empty()) {
             // For empty narrows where there's a clear reply target,
@@ -42,8 +47,8 @@ export function get_recipient_label(message) {
     if (message) {
         if (message.stream && message.topic) {
             return "#" + message.stream + " > " + message.topic;
-        } else if (message.recipients) {
-            return message.recipients;
+        } else if (message.display_reply_to) {
+            return message.display_reply_to;
         }
     }
     return "";
diff --git a/static/js/recent_topics_ui.js b/static/js/recent_topics_ui.js
index f9bdae27b9..ae468157ab 100644
--- a/static/js/recent_topics_ui.js
+++ b/static/js/recent_topics_ui.js
@@ -184,11 +184,14 @@ function set_table_focus(row, col, using_keyboard) {
         }
     }
 
+    // TODO: This fake "message" object is designed to allow using the
+    // get_recipient_label helper inside compose_closed_ui. Surely
+    // there's a more readable way to write this code.
     const type = get_row_type(row);
     let message;
     if (type === "private") {
         message = {
-            recipients: $topic_row.find(".recent_topic_name a").text(),
+            display_reply_to: $topic_row.find(".recent_topic_name a").text(),
         };
     } else {
         message = {
```

The other thing is I noticed a bug where the CSS seems pretty borked after rebasing onto `upstream/main`; not sure whether there's something wrong with my development environment or this PR conflicts with something we merged recently, but it'd be great if you can try to reproduce and fix it if so.

![image](https://user-images.githubusercontent.com/2746074/195961476-aa048076-6efb-4b42-b3e2-eb5a14d3ab53.png)

madrix01 commented:
@timabbott I have pushed the updated code with the fix for above encountered bug.

timabbott commented on static/templates/recent_topics_filters.hbs:
```diff
@@ -1,4 +1,12 @@
 <button data-filter="all" type="button" class="btn btn-default btn-recent-filters">{{t 'All' }}</button>
+<button data-filter="include_private" type="button" class="btn btn-default btn-recent-filters {{#if is_spectator}}fake_disabled_button{{/if}}" role="checkbox" aria-checked="true">
+    {{#if (or filter_pm not is_spectator)}}
```
This should be:
```
-    {{#if (or filter_pm not is_spectator)}}
+    {{#if filter_pm}}
```

`filter_pm` is always false if a spectator, so it doesn't accomplish anything, and this tweak better matches the other filters which all have the same property.

timabbott commented on static/js/recent_topics_data.js:
```diff
@@ -1,21 +1,21 @@
+
+    // Initialize topic and pm data
+    // Key for private message is the user id's
+    // to whom the message is begin sent.
```
This comment should explain the encoding.

timabbott commented on static/js/recent_topics_util.js:
```diff
@@ -29,3 +29,12 @@ export function is_in_focus() {
+
+export function get_key_from_message(msg) {
+    if (msg.type === "private") {
+        return msg.to_user_ids;
```
Added this explanatory comment:

        // The to_user_ids field on a private message object is a                                        
        // string containing the user IDs involved in the message in                                     
        // sorted order.

timabbott commented on static/js/recent_topics_util.js:
```diff
@@ -29,3 +29,12 @@ export function is_in_focus() {
+    } else if (msg.type === "stream") {
+        return get_topic_key(msg.stream_id, msg.topic);
+    }
+    return "";
```
-    return "";
+    throw new Error(`Invalid message type ${msg.type}`);

Other message types should be impossible; added an exception here.

timabbott commented:
OK I squashed the commits (since the first one broke Recent Topics without the second one included) and pushed back here with the above small changes made. I'm comfortable merging this change at this point.

Following this change there's a handful of things we'll want to consider including in 6.0:
* Renaming "Recent topics" to "Recent conversations", and related documentation updates. This is already a release goal as #23132.
* [ ] Making clicking the unread counts for private messages support "Mark as read" like we do for stream messages. I see no reason why we don't do that today but it seems like it should be a quick follow-up that @madrix01 and @amanagr can make happen in the next few days. @alya can you open a release goal issue for this?
* [ ] Adding "Last active time" hover behavior for the recent topics presence circles, similar to what you get hovering over presence circles in the app's sidebars. @alya can you open a release goal issue for this? 
* [ ] @madrix01 should open an issue for https://github.com/zulip/zulip/pull/20890#discussion_r997452326; I don't think it's a 6.0 release goal but I'd integrate a clean up if we can make it happen. #23250 is a similar category of non-urgent code quality issue that's apparent following this change (though that one is preexisting).
* [ ] https://github.com/zulip/zulip/pull/20890#discussion_r997441225 should generate a #frontend threat that we aim to conclude before 6.0 is released; I'm not familiar enough with this CSS feature to answer.

alya commented:
Sure, issues linked above. Thanks for the detailed summary!
